test/integration.js

```diff
@@ -0,0 +1,14 @@
+import Spark from '../lib/spark.js'
+import { test } from 'zinnia:test'
+
+test('integration', async () => {
+  const spark = new Spark()
+  for (let i = 0; i < 10; i++) {
+    try {
+      await spark.nextRetrieval()
+      break
+    } catch (err) {
+      console.error(err)
+    }
```

Comment:
Have you seen this test fail? It's not clear under which conditions we will get a test failure if you are converting errors to log messages.

Let's say we make this test as short and simple as possible:

```js
const spark = new Spark()
await spark.nextRetrieval()
```

What can go wrong?

Oh, I think I understand now. Are running up to 10 retrieval checks to handle the case when some of the content cannot be retrieved? If that's the case, then please add a code comment to explain. However, if all 10 retrieval checks fail, then the test is not going to fail, right? It will print 10 errors and then pass.

I also think that failed retrievals should be handled inside `nextRetrieval`, therefore we should not need to implement a retry scheme in this test.

Last but not least - it would be great if `nextRetrieval` returned retrieval `id` and this test could look up the retrieval record for that `id` using Fetch API, to verify that `nextRetrieval` performed some activity and reported the outcome. (Yes, I realise this requires changed in spark-api.)

Reply:
> > I also think that failed retrievals should be handled inside nextRetrieval, therefore we should not need to implement a retry scheme in this test.
> 
> Why do you think so? Failed retrievals are handled inside `Spark#run()`, which itself calls `Spark#nextRetrieval`, and handles failures thereof. Can you please explain why a different layout would be better?

Well, this is what I see inside `Spark#nextRetrieval`:

```js
    const retrieval = await this.getRetrieval()

    // ...
    try {
      stats = await this.fetchCAR(url)
      success = true
      this.#activity.onSuccess()
    } catch (err) {
      console.error(`Failed to fetch ${url}`)
      console.error(err)
      this.#activity.onError()
    }

    await this.submitRetrieval(retrieval.id, { success, ...stats })
    Zinnia.jobCompleted()
```

As I understand the code, `Spark#nextRetrieval` throws an error if we cannot get retrieval (something's wrong with Spark API) or submit retrieval result (something's wrong with Spark API). A failed retrieval attempt is converted to an error activity.

If my understanding is correct, then I am concurring that repeatedly calling `Spark#nextRetrieval` until it succeeds is something that deals with problems with Spark API. IMO, if Spark API is failing frequently enough that we need to handle that in our Spark Module test suite, then it's a major problem we should fix in Spark API.

If Spark API works reliably, it's enough to call `Spark#nextRetrieval` only once in the test.


> > Last but not least - it would be great if nextRetrieval returned retrieval id and this test could look up the retrieval record for that id using Fetch API, to verify that nextRetrieval performed some activity and reported the outcome. (Yes, I realise this requires changed in spark-api.)
> 
> I see your point, but I don't think we need this yet. Yes we can add a route to `spark-api` in order to help `spark`'s test suite, but at this stage I believe this level of control isn't yet beneficial. To me both `spark` and `spark-api` are still moving targets, and being able to move quickly is still of value.

My main concern is about subtle bugs that will make our test always pass, without invoking all components we want to test. This is very difficult to detect. If it happens, we will have a major bug and green CI checks.

I agree we should make it easy to iterate on Spark API and Spark Module independently of each other. At the same time, I think we can keep that property even with a new route to help with testing. All we need is a route that returns a boolean field telling us whether results were submitted for the given retrieval jobs. It can be as simple as using HTTP status codes (200 vs 404) to allow the test suite to tell whether results were reported or not.

Reply:
> If Spark API works reliably, it's enough to call Spark#nextRetrieval only once in the test.

I see! Thanks for the explanation. I had the error semantics messed up. You're right that we only need to run the function once. I have now fixed the function and the test.

> My main concern is about subtle bugs that will make our test always pass, without invoking all components we want to test. This is very difficult to detect. If it happens, we will have a major bug and green CI checks.

If we test that `spark` submitted the right data to `spark-api`, and the api responded with `200`, doesn't there the responsibility of `spark` end? Everything after that point I think should be covered by the `spark-api` test suite.

Nevertheless, it doesn't harm to verify this either, so I've created a PR to add that route: https://github.com/filecoin-station/spark-api/pull/21

Reply:
> > If Spark API works reliably, it's enough to call Spark#nextRetrieval only once in the test.
> 
> I see! Thanks for the explanation. I had the error semantics messed up. You're right that we only need to run the function once. I have now fixed the function and the test.

You are welcome! I am glad we are on the same page here 😅 

> > My main concern is about subtle bugs that will make our test always pass, without invoking all components we want to test. This is very difficult to detect. If it happens, we will have a major bug and green CI checks.
> 
> If we test that `spark` submitted the right data to `spark-api`, and the api responded with `200`, doesn't there the responsibility of `spark` end? Everything after that point I think should be covered by the `spark-api` test suite.

Right, I agree our responsibility ends when Spark API returns 200. The thing is, how can our test verify that Spark Module called Spark API and Spark API returned 200, when we are treating Spark Module as a black box and don't make any assumptions about its internals? In that case, we need to observe the entire system from the outside, and thus we need to check what was written to the database.

> Nevertheless, it doesn't harm to verify this either, so I've created a PR to add that route: [filecoin-station/spark-api#21](https://github.com/filecoin-station/spark-api/pull/21) I went with a more full featured response, which will help us debug retrievals easily.

🙏🏻

Reply:
Added the GET part to the integration test :)